symengine/tests/basic/test_solve.cpp

```diff
@@ -0,0 +1,315 @@
+
+    soln = solve(add({sqx, mul(i3, x), c}), x);
+    REQUIRE(soln->__str__()
+            == "{-3/2 + (-1/2)*sqrt(9 - 4*c), -3/2 + (1/2)*sqrt(9 - 4*c)}");
+
+    soln = solve(add({sqx, mul({i3, b, x}), c}), x);
+    REQUIRE(soln->__str__() == "{(-3/2)*b + (-1/2)*sqrt(-4*c + 9*b**2), "
+                               "(-3/2)*b + (1/2)*sqrt(-4*c + 9*b**2)}");
+
+    CHECK_THROWS_AS(solve_poly_quadratic({one}, x, reals), SymEngineException);
+
+    auto onebyx = div(one, x);
+    poly = add(onebyx, one);
+    soln = solve(poly, x);
+    REQUIRE(eq(*soln, *finiteset({neg(one)})));
+
+    poly = mul(onebyx, onebyx);
+    soln = solve(poly, x);
+    REQUIRE(eq(*soln, *emptyset()));
+}
+
+TEST_CASE("cubic and quartic polynomials", "[Solve]")
+{
+    RCP<const Symbol> x = symbol("x");
+    RCP<const Basic> poly;
+    RCP<const Set> reals = interval(NegInf, Inf, true, true);
+    RCP<const Set> soln;
+
+    auto sqx = mul(x, x), cbx = mul(sqx, x), qx = mul(cbx, x);
+    auto i2 = integer(2), i3 = integer(3), im2 = integer(-2);
+
+    // cubic
+    poly = Eq(add({cbx, mul(x, i3), mul(sqx, i3)}), neg(one));
+    soln = solve(poly, x, reals);
+    REQUIRE(eq(*soln, *finiteset({neg(one)})));
+
+    poly = Ne(add({cbx, mul(x, i3), mul(sqx, i3)}), neg(one));
+    soln = solve(poly, x, reals);
+    REQUIRE(eq(*soln, *SymEngine::set_union(
+                          {interval(NegInf, integer(-1), true, true),
+                           interval(integer(-1), Inf, true, true)})));
+
+    poly = Ge(add({cbx, mul(x, i3), mul(sqx, i3)}), one);
+    soln = solve(poly, x, reals);
+    REQUIRE(is_a<ConditionSet>(*soln));
+
+    REQUIRE(eq(*down_cast<const ConditionSet &>(*soln).get_condition(),
+               *logical_and({Ge(add({cbx, mul(x, i3), mul(sqx, i3)}), one),
+                             reals->contains(x)})));
+
+    poly = mul(cbx, i3);
+    soln = solve(poly, x, reals);
+    REQUIRE(eq(*soln, *finiteset({zero})));
+
+    poly = add(cbx, sub(add(mul(sqx, i3), mul(i3, x)), one));
+    soln = solve(poly, x);
+    auto r1 = neg(div(add(i3, pow(integer(-54), div(one, i3))), i3));
+    auto r2 = neg(div(add(i3, mul(add(div(one, im2), div(mul(I, sqrt(i3)), i2)),
+                                  pow(integer(-54), div(one, i3)))),
+                      i3));
+    auto r3 = neg(div(add(i3, mul(sub(div(one, im2), div(mul(I, sqrt(i3)), i2)),
+                                  pow(integer(-54), div(one, i3)))),
+                      i3));
+    REQUIRE(eq(*soln, *finiteset({r1, r2, r3})));
+
+    poly = sub(add(cbx, mul(integer(201), x)),
+               add(integer(288), mul(sqx, integer(38))));
+    soln = solve(poly, x);
+    r1 = i3;
+    r2 = integer(32);
+    REQUIRE(eq(*soln, *finiteset({r1, r2})));
+
+    poly = Eq(cbx, neg(one));
+    soln = solve(poly, x);
+    r1 = neg(one);
+    r2 = add(div(one, i2), div(mul(I, sqrt(i3)), i2));
+    r3 = sub(div(one, i2), div(mul(I, sqrt(i3)), i2));
+    // -(-1/2 - 1/2*I*sqrt(3)) != 1/2 + 1/2*I*sqrt(3) ?
```

Comment:
should we expand roots before returning so that roots like this `-(-1/2 - 1/2*I*sqrt(3))` is in a simplified form?
One disadvantage is that it will slow things down.